File: ocaml/bin/day03.ml
let read_input filename =
  let in_channel = open_in filename in
  let lines = In_channel.input_lines in_channel in
  close_in in_channel;

  List.fold_left (fun acc line -> acc ^ line) "" lines

let get_regex_matches input regex =
  let rec find_matches index acc =
    try
      let match_index = Str.search_forward regex input index in
      find_matches (match_index + 1) (acc @ [ Str.matched_string input ])
    with Not_found -> acc
  in
  find_matches 0 []

let extract_integers str = Scanf.sscanf str "mul(%d,%d)" (fun a b -> (a, b))

let extract_integers_and_mul str =
  let ints = extract_integers str in
  fst ints * snd ints

let silver () =
  let input = read_input "../inputs/03.txt" in
  let regex = Str.regexp "mul(\\([0-9]*\\),\\([0-9]*\\))" in
  let matches = get_regex_matches input regex in

  List.fold_left (fun a str -> a + extract_integers_and_mul str) 0 matches

let gold () =
  let input = read_input "../inputs/03.txt" in
  let regex = Str.regexp "mul(\\([0-9]*\\),\\([0-9]*\\))\\|do()\\|don't()" in
  let matches = get_regex_matches input regex in

  let rec summarize l enabled acc =
    match (l, enabled) with
    | [], _ -> acc
    | h :: t, _ when String.starts_with ~prefix:"don" h -> summarize t false acc
    | h :: t, _ when String.starts_with ~prefix:"do(" h -> summarize t true acc
    | h :: t, true -> summarize t true (acc + extract_integers_and_mul h)
    | _ :: t, false -> summarize t false acc
  in

  summarize matches true 0
;;

print_int (gold ())

File: cpp/day03.cpp
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <regex>
#include <string>
#include <vector>

std::string readInput() {
  std::string input;
  std::ifstream inputStream("../inputs/bigboy.txt");
  std::string line;

  while (std::getline(inputStream, line)) {
    input.append(line);
  }

  return input;
}

std::vector<std::string> getRegexMatches(const std::string &input,
                                         const std::regex &reg) {
  std::vector<std::string> res;

  std::smatch matches;
  std::regex_search(input, matches, reg);

  std::sregex_iterator regexIterator(input.begin(), input.end(), reg);
  std::sregex_iterator end;

  for (std::sregex_iterator k = regexIterator; k != end; ++k) {
    std::smatch match = *k;
    res.push_back(match.str());
  }

  return res;
}

int getMulResult(const std::string &mul) {
  int startIndex = mul.find('(');
  int endIndex = mul.find(",");
  int left = atoi(mul.substr(startIndex + 1, endIndex).c_str());
  startIndex = endIndex;
  endIndex = mul.find(')');
  int right = atoi(mul.substr(startIndex + 1, endIndex).c_str());

  return left * right;
}

int silver() {
  std::string input = readInput();
  std::regex reg("mul\\(\\d*,\\d*\\)");

  long sum = 0;
  std::vector<std::string> matches = getRegexMatches(input, reg);

  for (std::string match : matches) {
    sum += getMulResult(match);
  }

  return sum;
}

int gold() {
  std::string input = readInput();
  std::regex reg("mul\\(\\d*,\\d*\\)|do\\(\\)|don't\\(\\)");
  std::vector<std::string> matches = getRegexMatches(input, reg);

  bool enabled = true;
  long sum = 0;
  for (std::string match : matches) {
    if (match.substr(0, 3) == "do\(") {
      enabled = true;
      continue;
    } else if (match.substr(0, 3) == "don") {
      enabled = false;
      continue;
    }

    sum += enabled ? getMulResult(match) : 0;
  }

  return sum;
}

int main() {
  std::cout << "silver: " << silver() << "\n";
  std::cout << "gold: " << gold() << "\n";
}
